### ISSUE TITLE
focus.tabbables fails when the tabbables collection is made of only radio buttons

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#6988]
Splitting this out from #6987 

There's a problem with the `focus.tabbables` implementation when the tabbables are exclusively radio buttons. This is clearly visible in the "Visibility setting" when switching to the #6987 branch (see also screenshot below).

Radio buttons are technically "tabbables" but in a set of grouped radio buttons it's actually possible to tab into just one of them. In this case, the current implementation fails as `focus.tabbables` returns a collection of 3 elements but only one of them can be "tabbed" into:

<img width="319" alt="screen shot 2018-05-28 at 17 17 14" src="https://user-images.githubusercontent.com/1682452/40621527-65deca8e-629d-11e8-9666-00d235fafa32.png">

While technically all the radio buttons are tabbable, the current behavior doesn't help so much and in some cases it defeats the purpose of having such a tool. Gutenberg uses `focus.tabbables` mainly to get the next or previous tabbable elements and manage focus accordingly.

In the above scenario, the three radio buttons should be considered 1 tabbable element, to match the native browsers behavior.

**Comment by aduth** [comment]
A challenge here is directionality. For example, in the `WritingFlow` we move to the last tabbable when tabbing in reverse. `focus.tabbable` doesn't consider direction, so if we collapsed sequences of radio inputs to a single one (presumably the first), then shift-tabbing to a radio group would incorrectly move focus to the _first_ radio button, instead of the last in the set. 

Demo / Playground: https://codepen.io/aduth/pen/pKQKvm

**Comment by aduth** [comment]
One possibility: We add a `nextTabbable( node )` / `previousTabbable( node )` utility into the module to handle the directionality behaviors. Unsure how much of this could be reused from [`WritingFlow`'s `getClosestTabable`](https://github.com/WordPress/gutenberg/blob/3fd1328cac610cf3385a3d5db8dca894cac865f3/editor/components/writing-flow/index.js#L73-L145) which has additional logic to "prefer" fields instead of the block's focus stop. Perhaps `isTabCandidate` can be an argument to this utility function.

**Comment by afercia** [comment]
To clarify, the practical consequence of this in the Visibility panel is that tabbing with the keyboard is _not_ constrained within the panel. For example:
- when "Public" is selected, `focus.tabbables` thinks there are 3 tabbable elements
- the browser disagrees, as there's just one radio button which is "tabbable" at a time

So this is a case where constraining tabbing fails and screen reader users get lost.

Second example:
- select the last radio button "Password Protected"
- and additional input field for the password appears
- tab with the keyboard
- tabbing is now constrained but: it goes from the input to the _first_ radio button, which is not the one that should be tabbable

**Comment by aduth** [comment]
> In the above scenario, the three radio buttons should be considered 1 tabbable element, to match the native browsers behavior.

Considering the aforementioned complexities of directionality, would it be considered an improvement if tabbables considered all radio elements of the same name as a single element, where the chosen element is either the checked input (if available) or otherwise the first in the set?

It's non-perfect in that in a browser, Shift+Tab to focus into a radio group where there is no checked input would normally select the last radio in the DOM preceding the current focused element. But it would be relatively more trivial to implement this way.

Form with checked radio group: https://codepen.io/aduth/pen/xBbWrm
Form without checked radio group: https://codepen.io/aduth/pen/LaEdLv

I've sifted through a few relevant specifications ([[1]](https://www.w3.org/TR/html5/editing.html#sequential-focus-navigation) [[2]](https://html.spec.whatwg.org/multipage/input.html#radio-button-group)) but am yet to find where this particular behavior is prescribed. If it's not specified, it wouldn't surprise me then if this behavior differed across browsers.

**Comment by afercia** [comment]
Thanks for the codepens, really useful. Yep in the "without checked radio group" the behavior is different in Chrome and Firefox. On the other hand, a group of radio buttons should always have one of them selected. As you pointed out, in that case the expected behavior is unspecified and left to browsers implementations so... it differs 🙂 It's also a not ideal UI so that should be avoided.

Ideally, the unselected state of a radio buttons group shouldn't be allowed in Gutenberg, at least in the documentation.

That said, I'd tend to think your proposal is the best option available +1 !

**Comment by aduth** [comment]
From an already-interesting [Focus Management RFC](https://github.com/reactjs/rfcs/issues/104), I was led to the following recommendation:

>The convention for where focus lands in a composite when it receives focus as a result of a `Tab` key event depends on the type of composite. It is typically one of the following.
>
>- The element that had focus the last time the composite contained focus. Or, if the composite has not yet contained the focus, the first element. Widgets that usually employ this pattern include grid and tree grid.
>- The selected element. Or, if there is no selected element, the first element. Widgets where this pattern is commonly implemented include radio groups, tabs, list boxes, and trees. Note: For radio groups, this pattern is referring to the checked radio button; the selected state is not supported for radio buttons.
>- The first element. Components that typically follow this pattern include menubars and toolbars.

https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_general_within

The proposal in #14128 abides by the second and third, but does not consider the first of these points.

**Comment by afercia** [comment]
Interesting. From the comments on the same React RFC proposal:

> Focus management is there to repair the standard DOM focus when we disrupt it with our code.

Yes. 🙂 

Also, a very interesting consideration about portals, focus, and tab order (this was discussed at lenght for the popovers): https://github.com/reactjs/rfcs/issues/104#issuecomment-467110260

Besides that, worth noting the ARIA Authoring Practices relates to ARIA widgets. The Section "Keyboard Navigation Inside Components" relates to custom components built with ARIA and doesn't directly applies to _native_ elements and how browsers implement them.

The closest thing I've found about _native_ radio buttons is from the old HTML 4.0.1  recommendation:

https://www.w3.org/TR/html401/interact/forms.html#h-17.2.1

> If no radio button in a set sharing the same control name is initially "on", user agent behavior for choosing which control is initially "on" **is undefined**. Note. Since existing implementations handle this case differently, the current specification differs from RFC 1866 ([RFC1866] section 8.1.2.4), which states:

> > At all times, exactly one of the radio buttons in a set is checked. If none of the <INPUT> elements of a set of radio buttons specifies `CHECKED', then the user agent must check the first radio button of the set initially.

> Since user agent behavior differs, authors should ensure that in each set of radio buttons that one is initially "on".

The above basically just acknowledged browsers have different implementations. That said, #14128 is the best option!
